### PR TITLE
Use dependency injection inside controllers

### PR DESCRIPTION
**Pull request by FabienPapet**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | See below
| Type?             | new feature
| Category?         | CO
| BC breaks?        | yes 
| Deprecations?     | yes
| How to test?      | CI 🟢 and automated tests ~https://github.com/lartist/ga.tests.ui.pr/actions/runs/5070916874~ https://github.com/FabienPapet/ga.tests.ui.pr/actions/runs/5111658954 + see below

### Description

By default, Symfony provides a lightweight container that can be used inside controllers with some default services like twig, the form factory, etc. Also by default in Symfony, all controllers are services, which means they can have their dependencies injected inside the constructor for example.

There is a compiler pass in PrestaShop that has been included that removes that behavior and adds instead the full Dependency Injection Container (DIC) into the controllers, which forces us to mark the services as `public: true` if they are used inside a controller, which does not follow the Symfony standards and can lead to performance issues, as it avoids Symfony’s caching mechanism, that checks unused private services to remove them. This mechanism cannot be done on public services.

This PR is deprecating the `FrameworkBundleAdminController` and is introducing a new `PrestaShopAdminController` class that will respect Symfony guidelines, with autowiring and autoconfiguration enabled by default.

This adds many possibilities, 

- Some widely used services like configuration or the translator can now be injected within a small container that Symfony provides for the controllers (with twig for example).
- We can add some shortcuts to all controllers, to fetch grids in one line for example.
- It allows having services injected the Symfony way (no more `public: true` needed)

### Known Impacts

No impacts, `FrameworkBundleAdminControllers` will still continue to work as it used to.

## Testing

There are 2 controllers that ahve been migrated already.

- The security controller can be tested with a wrong token. We just test that the page is displayed without errors, the error that explains that the token is invalid is expected.
- The error controller page can be tested by enabling/disabling the debug mode.

### CODE REVIEW

**Review by tleon — DISMISSED** [review]

File: src/PrestaShopBundle/Resources/config/services/bundle/controller.yml [hunk 1]
```diff
@@ -2,30 +2,27 @@ services:
...
-    autoconfigure: true
+      - "%kernel.root_dir%/../src/PrestaShopBundle/Controller/ArgumentResolver"
+      - "%kernel.root_dir%/../src/PrestaShopBundle/Controller/Exception"
     autowire: true
+    autoconfigure: true
+    calls:
+      - [ setGlobalContainer, [ '@service_container' ] ]
     tags:
-      - { name: !php/const PrestaShopBundle\Controller\Admin\FrameworkBundleAdminController::PRESTASHOP_CORE_CONTROLLERS_TAG }
+      - 'controller.service_arguments'
```
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/bundle/controller.yml]: Is this manual tag needed if we already have `autoconfigure: true` ?

Comment (FabienPapet) [reply src/PrestaShopBundle/Resources/config/services/bundle/controller.yml]: You are right 👍

**Review by lartist — DISMISSED** [review]

**Review by M0rgan01 — DISMISSED** [review]

File: tests/Resources/config/services.yml [hunk 2]
```diff
@@ -44,3 +44,8 @@ services:
           - 'AdminLogin'
           - 'AdminQuickAccesses'
           - 'AdminTabs'
+
+  # This service is private and cannot be overriden in a test env, therefore we must set it public.
+    PrestaShop\PrestaShop\Core\Domain\Configuration\ShopConfigurationInterface:
+      class: PrestaShop\PrestaShop\Adapter\Configuration
+      public: true
```
Comment (lartist) [inline tests/Resources/config/services.yml:51]: Does it need to be public ?

Comment (M0rgan01) [reply tests/Resources/config/services.yml:51]: I think there is the answer in the code comment

**Review by lartist — DISMISSED** [review]

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/FeatureController.php [hunk 3]
```diff
@@ -53,6 +54,11 @@ class FeatureController extends FrameworkBundleAdminController
 {
     use BulkActionsTrait;
 
+    public function __construct(
+        private readonly FeatureFeature $feature
```
Comment (nicosomb) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/FeatureController.php]: 🤩

**Review by nicosomb — DISMISSED** [review]
OK for me. 
One failing test, is that normal?

**Comment by FabienPapet** [comment]
ping @PrestaShop/qa-automation for failing tests check

**Review by boherm — DISMISSED** [review]

**Comment by nesrineabdmouleh** [comment]
Hello @FabienPapet,

There are 2 :red_circle:  campaigns in UI tests :
- BO > Header : it's a regression in develop branch https://github.com/PrestaShop/PrestaShop/issues/32734
- BO > Design : error in design > link list page

Capture d’écran de 2023-05-30 10-05-49

**Comment by FabienPapet** [comment]
Failing test auto needs to be fixed, removing waiting for qa

**Comment by FabienPapet** [comment]
Should be all good now

**Review by boherm — DISMISSED** [review]

File: src/PrestaShopBundle/Controller/Admin/FrameworkBundleAdminController.php [hunk 4]
```diff
@@ -43,10 +45,27 @@
...
+class FrameworkBundleAdminController extends AbstractController implements ContainerAwareInterface
 {
+    /**
+     * @deprecated since 9.0
+     */
     public const PRESTASHOP_CORE_CONTROLLERS_TAG = 'prestashop.core.controllers';
 
+    /**
+     * Override to make this compatible with the ContainerAware signature, content should be the same as in the abstract.
+     * Do not override this neither use this, it will be removed in next versions.
```
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/FrameworkBundleAdminController.php]:
```suggestion
     * Do not override this neither use this, it will be removed in next versions. This overridden method
     * along with the ContainerAwareInterface was added to skip the error sent by Symfony\Bundle\FrameworkBundle\Controller\ControllerResolver
     * that forces the controllers extending AbstractController to be defined as service subscriber.
     *
     * This method allows us to keep controllers based on FrameworkBundleAdminController from being
     * adapted. However, the core and the modules should stop using it in favor of PrestaShopAdminController
```

Comment (jolelievre) [reply src/PrestaShopBundle/Controller/Admin/FrameworkBundleAdminController.php]: We should give a bit more detail to prevent someone from modifying this, and to help understand the need of this "hack"

File: src/PrestaShopBundle/Controller/Admin/PrestaShopAdminController.php [hunk 5]
```diff
@@ -24,13 +24,30 @@
...
+
+use PrestaShop\PrestaShop\Core\CommandBus\CommandBusInterface;
+use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
 
 /**
- * @deprecated since 8.0 and will be removed in next major. Use PrestaShop\PrestaShop\Core\ActionBar\ActionsBarButton instead
+ * Default controller for PrestaShop admin pages.
  */
-class ActionsBarButton extends CoreActionsBarButton implements ActionsBarButtonInterface
+class PrestaShopAdminController extends AbstractController
```
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/PrestaShopAdminController.php:37]:
```suggestion
class PrestaShopAdminController extends PrestaShopController
```

Comment (jolelievre) [reply src/PrestaShopBundle/Controller/Admin/PrestaShopAdminController.php:37]: I think we should also have a more generic parent `PrestaShopController` that will also be used as the parent class for `PrestaShopFrontController` later, in the parent we will probably add some generic services like configuration, hook manager and other services to define (twig? doctrine entity manager? ...)

Comment (jolelievre) [reply src/PrestaShopBundle/Controller/Admin/PrestaShopAdminController.php:37]: The command bus you added here would be relevant in the generic parent one as well IMHO, so the admin controller would end up empty for now But we can add some admin only services later, like the grid factory you started to implement

File: src/PrestaShopBundle/DependencyInjection/Compiler/ContainerInjectionPass.php [hunk 6]
```diff
@@ -41,6 +40,8 @@
  * Controllers lose the ControllerAwareTrait capabilities when they are decorated.
  *
  * This pass injects the container into PrestaShop tagged controllers to overcome this issue.
+ *
+ * @deprecated since 9.0, to be removed in 10.0. Controller are now services and can use dependency injection.
```
Comment (jolelievre) [inline src/PrestaShopBundle/DependencyInjection/Compiler/ContainerInjectionPass.php]:
```suggestion
 * @deprecated since 9.0, to be removed in a future version once all controllers have been migrated to proper service controllers based on PrestaShopAdminController.
```

Comment (jolelievre) [reply src/PrestaShopBundle/DependencyInjection/Compiler/ContainerInjectionPass.php]: We can't be sure of when the full refacto will be finished so it's probably more relevant to indicate what needs to be achieved before this can be removed instead of a fixed version

File: src/PrestaShopBundle/Controller/Admin/PrestaShopAdminController.php
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/PrestaShopAdminController.php:1]: I don't understand why you remove these object action bars, it doesn't seem related with the PR's scope and it's still used in AdminController and in a hook https://github.com/PrestaShop/PrestaShop/blob/8af2365e821403d6dc7f8c6b03165600db4cd33f/classes/controller/AdminController.php#L4382

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/ActionsBarButtonInterface.php
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Sell/Order/ActionsBarButtonInterface.php:1]: This should be restored

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/ActionsBarButtonsCollection.php
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Sell/Order/ActionsBarButtonsCollection.php:1]: This one as well shouldn't be removed I think 🤔

File: src/PrestaShopBundle/Controller/Admin/Improve/ModuleController.php [hunk 7]
```diff
@@ -192,9 +192,9 @@ public function moduleAction(Request $request)
 
         $module = $request->get('module_name');
         $source = $request->query->get('source');
-        $moduleManager = $this->container->get('prestashop.module.manager');
```
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Improve/ModuleController.php:195]: Is this still relevant? Or is it a modification from your previous implementation? If it's not required maybe we should revert this and keep the PR as small as possible to only have code related to what we're aiming here

File: src/PrestaShopBundle/DependencyInjection/Compiler/ContainerInjectionPass.php [hunk 8]
```diff
@@ -66,7 +67,7 @@ public function process(ContainerBuilder $container)
 
             $isContainerAware = (
                 $reflectedClass->implementsInterface(ContainerAwareInterface::class)
-                || is_subclass_of($class, AbstractController::class)
+                || is_subclass_of($class, FrameworkBundleAdminController::class)
```
Comment (jolelievre) [inline src/PrestaShopBundle/DependencyInjection/Compiler/ContainerInjectionPass.php]: Since `FrameworkBundleAdminController` implements `ContainerAwareInterface` this second condition can be removed no? Unless we revert the condition and make `FrameworkBundleAdminController` the only condition when we force to inject the global container?

File: src/PrestaShopBundle/DependencyInjection/Compiler/LoadServicesFromModulesPass.php [hunk 9]
```diff
@@ -37,6 +37,8 @@
 
 /**
  * Load services stored in installed modules.
+ *
+ * @deprecated since 9.0, to be removed in 10.0
```
Comment (jolelievre) [inline src/PrestaShopBundle/DependencyInjection/Compiler/LoadServicesFromModulesPass.php]:
```suggestion
 * @deprecated since 9.0, not used anymore to be removed in 10.0
```

Comment (jolelievre) [reply src/PrestaShopBundle/DependencyInjection/Compiler/LoadServicesFromModulesPass.php]: I hesitate to remove this right away, it's not like it's a class that has chances to have been used or extended outside of the core 🤔

File: src/PrestaShopBundle/Resources/config/services/bundle/controller.yml [hunk 10]
```diff
@@ -2,29 +2,35 @@ services:
   _defaults:
     public: true
 
+  _instanceof:
+    # cannot use autocnfigure under _instanceof so we need to define it here
```
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/bundle/controller.yml]:
```suggestion
    # cannot use autoconfigure under _instanceof so we need to define it here
```

File: src/PrestaShopBundle/Resources/config/services/bundle/controller.yml [hunk 11]
```diff
@@ -2,29 +2,35 @@ services:
...
+    PrestaShopBundle\Controller\Admin\PrestaShopAdminController:
+      autowire: true
+      public: true
+      tags:
+        - 'controller.service_arguments'
+        - 'container.service_subscriber'
+    PrestaShopBundle\Controller\Admin\FrameworkBundleAdminController:
```
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/bundle/controller.yml]: If we have this do we still need the `ContainerInjectionPass` ? These two solutions seem redundant no?

Comment (jolelievre) [reply src/PrestaShopBundle/Resources/config/services/bundle/controller.yml]: If we go with this solution only we should comment here that this rule handles the fact that controllers based on `FrameworkBundleAdminController` are the only one where we still inject the full container

File: src/PrestaShopBundle/Resources/config/services/bundle/controller.yml [hunk 12]
```diff
@@ -2,29 +2,35 @@ services:
...
+
+
+  # Legacy controllers don't extend the base controller, so we need to inject the container manually.
   PrestaShopBundle\Controller\Api\ApiController:
     class: PrestaShopBundle\Controller\Api\ApiController
     calls:
       - [ setLogger, [ '@logger' ] ]
       - [ setContainer, [ '@service_container' ] ]
-    tags:
-      - { name: !php/const PrestaShopBundle\Controller\Admin\FrameworkBundleAdminController::PRESTASHOP_CORE_CONTROLLERS_TAG }
```
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/bundle/controller.yml:11]: The tag is not needed anymore? Then it means the `ContainerInjectionPass` doesn't find tagged services anymore since most of them don't have the tag anymore right?

File: src/PrestaShopBundle/Controller/Admin/Improve/ModuleController.php (hunk 7 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Controller/Admin/Improve/ModuleController.php:195]: After discussion, the purpose here is to harmonize our controllers and only use `$this->get` A PHPStan rule should be added to prevent his kind of usage later

File: src/PrestaShopBundle/Controller/Admin/FrameworkBundleAdminController.php [hunk 13]
```diff
@@ -43,10 +45,27 @@
 /**
  * Extends The Symfony framework bundle controller to add common functions for PrestaShop needs.
  */
-class FrameworkBundleAdminController extends AbstractController
+class FrameworkBundleAdminController extends AbstractController implements ContainerAwareInterface
```
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/FrameworkBundleAdminController.php:50]: Add deprecated notice (without fixed version for removal) to explain that we should favor PrestaShopAdminController instead. This class will be removed in a major version (at least 10) once all the usage of this class have been removed from the core

File: src/PrestaShopBundle/Controller/Admin/ErrorController.php [hunk 14]
```diff
@@ -34,7 +34,7 @@
 /**
  * Manages Error pages (e.g. 500)
  */
-class ErrorController extends FrameworkBundleAdminController
+class ErrorController extends PrestaShopAdminController
```
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/ErrorController.php:37]: This should be detailed in the How to test to know how (probably a developer) will have to test this PR

File: src/PrestaShopBundle/Resources/config/services/bundle/controller.yml (hunk 11 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Resources/config/services/bundle/controller.yml]: See if the modif in the class itself is enough, if yes then this should be cleaned

File: src/PrestaShopBundle/DependencyInjection/Compiler/ContainerInjectionPass.php
Comment (jolelievre) [inline src/PrestaShopBundle/DependencyInjection/Compiler/ContainerInjectionPass.php:1]: This file should be removed

File: src/PrestaShopBundle/DependencyInjection/Compiler/LoadServicesFromModulesPass.php
Comment (jolelievre) [inline src/PrestaShopBundle/DependencyInjection/Compiler/LoadServicesFromModulesPass.php:1]: This file should be removed

Comment (FabienPapet) [reply src/PrestaShopBundle/DependencyInjection/Compiler/LoadServicesFromModulesPass.php:1]: It is used by the front container and cannot be removed for now

File: src/PrestaShopBundle/DependencyInjection/Compiler/LoadServicesFromModulesPass.php (hunk 9 above)
Comment (jolelievre) [inline src/PrestaShopBundle/DependencyInjection/Compiler/LoadServicesFromModulesPass.php]:
```suggestion
 * @deprecated since 9.0, to be removed once it's not used by PrestaShop\PrestaShop\Adapter\ContainerBuilder
```

File: src/PrestaShopBundle/Controller/Admin/FrameworkBundleAdminController.php [hunk 15]
```diff
@@ -34,19 +34,45 @@
...
+use Symfony\Component\DependencyInjection\ContainerAwareInterface;
 use Symfony\Component\Form\FormError;
 use Symfony\Component\Form\FormInterface;
 use Symfony\Component\HttpFoundation\JsonResponse;
 use Symfony\Component\HttpFoundation\RedirectResponse;
 
 /**
  * Extends The Symfony framework bundle controller to add common functions for PrestaShop needs.
+ *
+ * @deprecated since 9.0 to be removed in future versions (10+ at least, when it will no more used)
```
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/FrameworkBundleAdminController.php]:
```suggestion
 * @deprecated since 9.0 to be removed in future versions (10+ at least, when it will not be used anymore)
```

File: src/PrestaShopBundle/Controller/Admin/PrestaShopAdminController.php
Comment (FabienPapet) [reply src/PrestaShopBundle/Controller/Admin/PrestaShopAdminController.php:1]: https://github.com/PrestaShop/PrestaShop/pull/32927

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/ActionsBarButtonsCollection.php
Comment (FabienPapet) [reply src/PrestaShopBundle/Controller/Admin/Sell/Order/ActionsBarButtonsCollection.php:1]: https://github.com/PrestaShop/PrestaShop/pull/32927

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/ActionsBarButtonInterface.php
Comment (FabienPapet) [reply src/PrestaShopBundle/Controller/Admin/Sell/Order/ActionsBarButtonInterface.php:1]: https://github.com/PrestaShop/PrestaShop/pull/32927

File: app/AppKernel.php [hunk 16]
```diff
@@ -204,24 +204,28 @@ public function getLogDir()
...
-            foreach ($this->getActiveModules() as $activeModulePath) {
-                $translationsDir = _PS_MODULE_DIR_ . $activeModulePath . '/translations';
-                if (is_dir($translationsDir)) {
-                    $moduleTranslationsPaths[] = $translationsDir;
-                }
+        $moduleTranslationsPaths = [];
+        foreach ($installedModules as $activeModulePath)
+        {
+            $modulePath = _PS_MODULE_DIR_ . $activeModulePath;
+            $configPath = sprintf('%s/config/services.yml', $modulePath);
```
Comment (jolelievre) [inline app/AppKernel.php]: Do the same with `sprintf('%s/config/admin/services.yml', $modulePath);`

**Review by jolelievre — APPROVED** [review]
Thanks @FabienPapet

**Review by boherm — APPROVED** [review]

**Review by M0rgan01 — APPROVED** [review]

**Comment by jolelievre** [comment]
I tested the PR, both modified controllers behave as wanted And this PR removes a previously introduced breaking change (or at least reduces it) which allows us to revert a modification from ps_linklist https://github.com/PrestaShop/ps_linklist/pull/175

And here is a campaign to check all UI tests are fine https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/5331496441

**Comment by matthieu-rolland** [comment]
thank you @FabienPapet :+1: 

UI tests are all green, and @jolelievre validated the behavior, I merge the PR.
